astropy/coordinates/transformations.py

```diff
@@ -746,8 +749,157 @@ def __call__(self, fromcoord, toframe):
         if not isinstance(res, self.tosys):
             raise TypeError('the transformation function yielded {0} but '
                 'should have been of type {1}'.format(res, self.tosys))
+        if (getattr(fromcoord.data, 'differentials', None) and not
+            getattr(res.data, 'differentials', None)):
+            warn("Applied a FunctionTransform to a coordinate frame with "
+                 "differentials, but the FunctionTransform does not handle "
+                 "differentials, so they have been dropped.", AstropyWarning)
         return res
 
+class FunctionTransformWithFiniteDifference(FunctionTransform):
+    """
+    A coordinate transformation that works like a `FunctionTransform`, but
+    computes velocity shifts based on the finite-difference relative to one of
+    the frame attributes.  Note that the transform function should *not* change
+    the differential at all in this case, as any differentials will be
+    overridden.
+
+    Parameters
+    ----------
+    finite_difference_frameattr_name : str or None
+        The name of the frame attribute on the frames to use for the finite
+        difference.  Both the to and the from frame will be checked for this
+        attribute, but only one needs to have it. If None, only the
+        re-orientation of the existing differential will be computed.
```

Comment:
will try to do so - it's basically the matrix mart of the affine transform (applied to the velocities)